mmdet/apis/ote/apis/detection/nncf_task.py

```diff
@@ -0,0 +1,520 @@
+
+        self.task_environment = task_environment
+        self.hyperparams = hyperparams = task_environment.get_hyper_parameters(OTEDetectionConfig)
+
+        self.model_name = hyperparams.algo_backend.model_name
+        self.labels = task_environment.get_labels(False)
+
+        template_file_path = task_environment.model_template.model_template_path
+
+        # Get and prepare mmdet config.
+        base_dir = os.path.abspath(os.path.dirname(template_file_path))
+        config_file_path = os.path.join(base_dir, hyperparams.algo_backend.model)
+        self.config = Config.fromfile(config_file_path)
+        patch_config(self.config, self.scratch_space, self.labels, random_seed=42)
+        set_hyperparams(self.config, hyperparams)
+
+        # NNCF part
+        self.compression_ctrl = None
+        nncf_config_path = os.path.join(base_dir, hyperparams.nncf_optimization.config)
+
+        with open(nncf_config_path) as nncf_config_file:
+            common_nncf_config = json.load(nncf_config_file)
+
+
+        nncf_config = common_nncf_config["base"]["nncf_config"]
+        optimization_type = None
+        if hyperparams.nncf_optimization.apply_quantization:
+            optimization_type = "nncf_quantization"
+        elif hyperparams.nncf_optimization.apply_pruning:
+            optimization_type = "nncf_pruning"
+        elif hyperparams.nncf_optimization.apply_pruning_quantization:
+            optimization_type = "nncf_pruning_quantization"
+        else:
+            raise ValueError("Optimization was disabled")
+
+
+        compesed = compose_nncf_config(common_nncf_config, [optimization_type])
+        self.config.update(compesed)
+
+        # Create and initialize PyTorch model.
+        check_nncf_is_enabled()
+        self.compression_ctrl, self.model = self._load_model(task_environment.model)
+
+        # Extra control variables.
+        self.training_work_dir = None
+        self.is_training = False
+        self.should_stop = False
+        self.time_monitor = None
+
+    @staticmethod
+    def extract_model_and_compression_states(resuming_checkpoint: Optional[Dict] = None):
+        if resuming_checkpoint is None:
+            return None, None
+        model_state_dict = resuming_checkpoint.get("model" if "model" in resuming_checkpoint else "state_dict")
+        compression_state = resuming_checkpoint.get("compression_state")
+        return model_state_dict, compression_state
+
+    def _load_model(self, model: Model):
+        if model != NullModel():
+            # If a model has been trained and saved for the task already, create empty model and load weights here
+            buffer = io.BytesIO(model.get_data("weights.pth"))
+            cfgs = model
+            model_data = torch.load(buffer, map_location=torch.device('cpu'))
+
+            model = self._create_model(self.config, from_scratch=True)
+            compression_ctrl = None
+            try:
+                if 'compression_state' in model_data:
```

Comment:
The loading logic should be different. We should check whether the model is already NNCF or not, wrap the model in any case, and adjust parameters restoration accordingly. @alexsu52, please take a look.

Reply:
it's works like you describe, but one monent. In case of training from non-nncf weights we can not wrap model in __init__ because we have no dataset to initialize NNCFNetwork as result for this case wrapping model move to optimize function (on init build mmdetection model only)